File: week09/26/prob1.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    bool dfs(int curr, int destination, vector<vector<int>>& adj, vector<bool>& visited){
        if(curr == destination) {
            return true;
        }
        visited[curr] = true;
        for(int i = 0; i < adj[curr].size(); ++i){
            int neighbor = adj[curr][i];
            if(!visited[neighbor]){
                if(dfs(neighbor, destination, adj, visited)) {
                    return true;
                }
            }
        }
        return false;
    }
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        vector<vector<int>> adj(n, vector<int>());
        for(int i=0; i<edges.size(); ++i){
            vector<int> e = edges[i];
            adj[e[0]].push_back(e[1]);
            adj[e[1]].push_back(e[0]);
        }

        vector<bool> visited(n, 0);
        return dfs(source, destination, adj, visited);
    }
    
    
};

File: week09/26/prob1.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        adj = collections.defaultdict(list)
        for s, d in edges:
            adj[s] += [d]
            adj[d] += [s]



        if len(adj) == 0:
            return True

        to_visit = []
        visited = []
        to_visit += adj[source]
        to_visit = deque(to_visit)
        visited = set(visited)

        while to_visit:
            curr = to_visit.pop()
            if curr == destination:
                return True
            for neighbor in adj[curr]:
                if neighbor not in visited:
                    visited.add(neighbor)
                    to_visit += [neighbor]
                
        return False

